stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,238 @@
+    scales than those automatically calculated.
+    """
+
+    def __init__(self, G, scales, num_eigenvecs=-1):
+        """
+        Args:
+            G (StellarGraph): the StellarGraph object.
+            scales (str or list of floats): the wavelet scales to use. "auto" will cause the scale values to be
+                automatically calculated.
+            num_scales (int): the number of scales when scales = "auto".
+            num_eigenvecs (int): the number of eigenvectors to use. When set to `-1` the number of eigenvectors
+                is automatically determined.
+        """
+
+        if not isinstance(G, StellarGraph):
+            raise TypeError("G must be a StellarGraph object.")
+
+        node_types = list(G.node_types)
+        if len(node_types) > 1:
+            raise TypeError(
+                "{}: node generator requires graph with single node type; "
+                "a graph with multiple node types is passed. Stopping.".format(
+                    type(self).__name__
+                )
+            )
+
+        # Create sparse adjacency matrix:
+        adj = G.to_adjacency_matrix().tocoo()
+
+        # Function to map node IDs to indices for quicker node index lookups
+        self._node_lookup = G._get_index_for_nodes
+
+        degree_mat = diags(np.asarray(adj.sum(1)).ravel())
+        laplacian = degree_mat - adj
+
+        # TODO: add in option to compute wavelet transform using Chebysev polynomials
+
+        self.scales = np.array(scales).astype(np.float32)
+
+        if num_eigenvecs == -1:
+            eigen_vals, eigen_vecs = self.sufficiently_sampled_eigs(
+                laplacian, self.scales.min()
+            )
+        else:
+            eigen_vals, eigen_vecs = eigs(laplacian, k=num_eigenvecs, sigma=1e-3)
+            eigen_vals = np.real(eigen_vals).astype(np.float32)
+            eigen_vecs = eigen_vecs[:, 1:]
+            eigen_vals = eigen_vals[1:]
+
+        self.eigen_vecs = np.real(eigen_vecs).astype(np.float32)
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of U exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+
+        # a list of [U exp(-scale * eigenvalues) for scale in scales]
+        Ues = [
+            (self.eigen_vecs * np.exp(-s * eigen_vals))[np.newaxis, :, :].astype(
+                np.float32
+            )
+            for s in scales
+        ]
+        self.Ues = tf.convert_to_tensor(np.concatenate(Ues, axis=0))
+
+    @staticmethod
+    def sufficiently_sampled_eigs(laplacian, min_scale):
```

Comment:
I suspect a user shouldn't ever call this, and listing it in the docs will be a distraction, so maybe:

```suggestion
    def _sufficiently_sampled_eigs(laplacian, min_scale):
```